mackerel-plugin-aws-kinesis-firehose/lib/aws-kinesis-firehose.go

```diff
@@ -0,0 +1,243 @@
+package mpawskinesisfirehose
+
+import (
+	"errors"
+	"flag"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	mp "github.com/mackerelio/go-mackerel-plugin-helper"
+)
+
+const (
+	namespace = "AWS/Firehose"
+)
+
+type metrics struct {
+	CloudWatchName string
+	MackerelName   string
+}
+
+// KinesisFirehosePlugin mackerel plugin for aws kinesis firehose
+type KinesisFirehosePlugin struct {
+	Name   string
+	Prefix string
+
+	AccessKeyID     string
+	SecretAccessKey string
+	Region          string
+	CloudWatch      *cloudwatch.CloudWatch
+}
+
+// MetricKeyPrefix interface for PluginWithPrefix
+func (p KinesisFirehosePlugin) MetricKeyPrefix() string {
+	if p.Prefix == "" {
+		p.Prefix = "kinesis-firehose"
+	}
+	return p.Prefix
+}
+
+func (p *KinesisFirehosePlugin) prepare() error {
+	sess, err := session.NewSession()
+	if err != nil {
+		return err
+	}
+
+	config := aws.NewConfig()
+	if p.AccessKeyID != "" && p.SecretAccessKey != "" {
+		config = config.WithCredentials(credentials.NewStaticCredentials(p.AccessKeyID, p.SecretAccessKey, ""))
+	}
+	if p.Region != "" {
+		config = config.WithRegion(p.Region)
+	}
+
+	p.CloudWatch = cloudwatch.New(sess, config)
+
+	return nil
+}
+
+// getLastPoint fetches a CloudWatch metric and parse
+func (p KinesisFirehosePlugin) getLastPoint(metric metrics) (float64, error) {
+	now := time.Now()
+
+	dimensions := []*cloudwatch.Dimension{
+		{
+			Name:  aws.String("DeliveryStreamName"),
+			Value: aws.String(p.Name),
+		},
+	}
+
+	response, err := p.CloudWatch.GetMetricStatistics(&cloudwatch.GetMetricStatisticsInput{
+		Dimensions: dimensions,
+		StartTime:  aws.Time(now.Add(time.Duration(180) * time.Second * -1)), // 3 min
+		EndTime:    aws.Time(now),
+		MetricName: aws.String(metric.CloudWatchName),
+		Period:     aws.Int64(60),
+		Statistics: []*string{aws.String("Average")},
+		Namespace:  aws.String(namespace),
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	datapoints := response.Datapoints
+	if len(datapoints) == 0 {
+		return 0, errors.New("fetched no datapoints")
```

Comment:
Thank you for the review.

> According to document, CloudWatch metrics of Firehose is collected every 5 minutes.
> Therefore, this plugin may fetch no datapoints even if CloudWatch and plugin are working correctly.

I overeloooked the document. I changed duration time. 

> Currently mackerel-agent outputs all logs written by plugins, so with this implementation, many "no datapoints" logs by this plugin will be written to mackerel-agent.log.
> Do you have any idea to suppress this error?

I suppressed "no datapoints" logs.